File: project-{topic}-plus/app/services/user_service.py
```python
from app.extensions import db
from app.models.user import User, UserRole
from app.schemas.user import user_register_schema, user_update_schema, user_schema
from app.utils.errors import NotFoundError, ConflictError, BadRequestError
from sqlalchemy.exc import IntegrityError
import logging

class UserService:
    """
    Service layer for User-related business logic.
    Handles CRUD operations and user-specific actions like password changes.
    """
    def __init__(self):
        self.logger = logging.getLogger(__name__)

    def get_all_users(self):
        """Retrieves all users."""
        self.logger.info("Fetching all users.")
        users = User.query.all()
        return users_schema.dump(users)

    def get_user_by_id(self, user_id):
        """Retrieves a single user by ID."""
        self.logger.info(f"Fetching user with ID: {user_id}")
        user = User.query.get(user_id)
        if not user:
            raise NotFoundError(f"User with ID {user_id} not found.")
        return user_schema.dump(user)

    def get_user_by_username(self, username):
        """Retrieves a single user by username."""
        self.logger.info(f"Fetching user with username: {username}")
        user = User.query.filter_by(username=username).first()
        if not user:
            raise NotFoundError(f"User with username '{username}' not found.")
        return user

    def create_user(self, user_data):
        """
        Creates a new user.
        Validates input and handles password hashing.
        """
        self.logger.info(f"Attempting to create user with data: {user_data.get('username')}")
        
        # Validate input using Marshmallow schema
        errors = user_register_schema.validate(user_data)
        if errors:
            raise BadRequestError(f"Validation errors: {errors}")

        # Check for existing username or email before attempting to add
        if User.query.filter_by(username=user_data['username']).first():
            raise ConflictError(f"Username '{user_data['username']}' already exists.")
        if User.query.filter_by(email=user_data['email']).first():
            raise ConflictError(f"Email '{user_data['email']}' already exists.")

        try:
            user = User(
                username=user_data['username'],
                email=user_data['email'],
                password=user_data['password'], # Password will be hashed by model's constructor
                role=UserRole(user_data.get('role', UserRole.USER.value)) # Default to USER
            )
            db.session.add(user)
            db.session.commit()
            self.logger.info(f"User '{user.username}' created successfully with ID: {user.id}")
            return user_schema.dump(user)
        except IntegrityError as e:
            db.session.rollback()
            self.logger.error(f"Error creating user: {e}", exc_info=True)
            if "duplicate key value violates unique constraint" in str(e):
                raise ConflictError("A user with this username or email already exists.")
            raise

    def update_user(self, user_id, update_data):
        """
        Updates an existing user's information.
        Handles password changes if 'new_password' is provided.
        """
        self.logger.info(f"Attempting to update user ID: {user_id} with data: {update_data}")
        user = User.query.get(user_id)
        if not user:
            raise NotFoundError(f"User with ID {user_id} not found.")

        # Validate input using Marshmallow schema for updates
        errors = user_update_schema.validate(update_data, partial=True)
        if errors:
            raise BadRequestError(f"Validation errors: {errors}")

        try:
            # Handle unique constraints for username and email
            if 'username' in update_data and update_data['username'] != user.username:
                if User.query.filter_by(username=update_data['username']).first():
                    raise ConflictError(f"Username '{update_data['username']}' already exists.")
                user.username = update_data['username']
            
            if 'email' in update_data and update_data['email'] != user.email:
                if User.query.filter_by(email=update_data['email']).first():
                    raise ConflictError(f"Email '{update_data['email']}' already exists.")
                user.email = update_data['email']

            if 'role' in update_data:
                user.role = UserRole(update_data['role'])
            
            if 'is_active' in update_data:
                user.is_active = update_data['is_active']
            
            # Handle password change
            if 'new_password' in update_data and update_data['new_password']:
                user.set_password(update_data['new_password'])

            db.session.commit()
            self.logger.info(f"User ID: {user.id} updated successfully.")
            return user_schema.dump(user)
        except IntegrityError as e:
            db.session.rollback()
            self.logger.error(f"Error updating user ID {user_id}: {e}", exc_info=True)
            if "duplicate key value violates unique constraint" in str(e):
                raise ConflictError("A user with this username or email already exists.")
            raise

    def delete_user(self, user_id):
        """Deletes a user by ID."""
        self.logger.info(f"Attempting to delete user ID: {user_id}")
        user = User.query.get(user_id)
        if not user:
            raise NotFoundError(f"User with ID {user_id} not found.")
        
        db.session.delete(user)
        db.session.commit()
        self.logger.info(f"User ID: {user.id} deleted successfully.")
        return {"message": f"User with ID {user_id} deleted successfully."}

```

File: project-{topic}-plus/src/services/userService.js
```javascript
const userRepository = require('../repositories/userRepository');
const AppError = require('../utils/appError');
const APIFeatures = require('../utils/apiFeatures');
const bcrypt = require('bcryptjs');

/**
 * User Service Module
 * Handles business logic for user management (excluding authentication).
 */
const userService = {
  /**
   * Retrieves a user by their ID.
   * @param {string} userId - The ID of the user to retrieve.
   * @returns {Promise<object>} The user object.
   * @throws {AppError} If the user is not found.
   */
  async getUserById(userId) {
    const user = await userRepository.findById(userId);
    if (!user) {
      throw new AppError('User not found.', 404);
    }
    return user;
  },

  /**
   * Updates a user's profile.
   * @param {string} userId - The ID of the user to update.
   * @param {object} updateData - The data to update (e.g., username, email).
   * @returns {Promise<object>} The updated user object.
   * @throws {AppError} If the user is not found or if email/username already exists.
   */
  async updateProfile(userId, updateData) {
    // 1. Check if user exists
    const user = await userRepository.findById(userId);
    if (!user) {
      throw new AppError('User not found.', 404);
    }

    // 2. Prevent updating role via this endpoint
    if (updateData.role) {
      throw new AppError('Cannot update user role via this endpoint.', 403); // 403 Forbidden
    }

    // 3. Check for duplicate email/username if provided
    if (updateData.email && updateData.email !== user.email) {
      const existingUser = await userRepository.findByEmail(updateData.email);
      if (existingUser) {
        throw new AppError('Email already registered.', 409);
      }
    }

    if (updateData.username && updateData.username !== user.username) {
      const existingUser = await userRepository.findByUsername(updateData.username);
      if (existingUser) {
        throw new AppError('Username already taken.', 409);
      }
    }

    // 4. Update user
    const updatedUser = await userRepository.update(userId, updateData);
    return updatedUser;
  },

  /**
   * Updates a user's password.
   * @param {string} userId - The ID of the user.
   * @param {string} currentPassword - The user's current password.
   * @param {string} newPassword - The new password.
   * @returns {Promise<object>} The updated user object (without password).
   * @throws {AppError} If user not found, current password incorrect, or new password is same as old.
   */
  async updatePassword(userId, currentPassword, newPassword) {
    // 1. Get user with password hash
    const user = await userRepository.findById(userId, true); // Need password for comparison
    if (!user) {
      throw new AppError('User not found.', 404);
    }

    // 2. Check if current password is correct
    const isPasswordCorrect = await bcrypt.compare(currentPassword, user.password);
    if (!isPasswordCorrect) {
      throw new AppError('Current password is incorrect.', 401);
    }

    // 3. Check if new password is different from current
    const isNewPasswordSame = await bcrypt.compare(newPassword, user.password);
    if (isNewPasswordSame) {
      throw new AppError('New password cannot be the same as the current password.', 400);
    }

    // 4. Hash new password and update
    const hashedPassword = await bcrypt.hash(newPassword, 12);
    const updatedUser = await userRepository.update(userId, { password: hashedPassword });

    return updatedUser;
  },

  /**
   * Deactivates a user account.
   * @param {string} userId - The ID of the user to deactivate.
   * @returns {Promise<void>}
   * @throws {AppError} If the user is not found.
   */
  async deactivateAccount(userId) {
    // In a real application, you might set a `isActive: false` flag
    // or a `deletedAt` timestamp instead of actual deletion for data retention.
    // For this example, we'll perform a soft-delete simulation by throwing an error
    // or a full delete for simplicity if desired.
    // Let's implement a full delete for simplicity but note the real-world implications.
    const user = await userRepository.findById(userId);
    if (!user) {
      throw new AppError('User not found.', 404);
    }
    await userRepository.delete(userId);
  },

  /**
   * Admin-only: Retrieves all users with pagination, filtering, and sorting.
   * @param {object} queryString - The query parameters from the request (req.query).
   * @returns {Promise<object>} An object containing results and total count.
   */
  async getAllUsers(queryString) {
    const features = new APIFeatures(userRepository.prisma.user, queryString)
      .filter()
      .sort()
      .limitFields()
      .paginate();

    const users = await features.execute();
    const totalCount = await features.count();

    return {
      results: users.length,
      total: totalCount,
      users,
    };
  },

  /**
   * Admin-only: Updates a user's role.
   * @param {string} userId - The ID of the user to update.
   * @param {string} role - The new role (e.g., 'ADMIN', 'USER').
   * @returns {Promise<object>} The updated user object.
   * @throws {AppError} If user not found or invalid role.
   */
  async updateUserRole(userId, role) {
    const user = await userRepository.findById(userId);
    if (!user) {
      throw new AppError('User not found.', 404);
    }

    if (!['USER', 'ADMIN'].includes(role.toUpperCase())) {
      throw new AppError('Invalid user role provided.', 400);
    }

    const updatedUser = await userRepository.update(userId, { role: role.toUpperCase() });
    return updatedUser;
  },

  /**
   * Admin-only: Deletes a user account.
   * @param {string} userId - The ID of the user to delete.
   * @returns {Promise<void>}
   * @throws {AppError} If the user is not found.
   */
  async deleteUser(userId) {
    const user = await userRepository.findById(userId);
    if (!user) {
      throw new AppError('User not found.', 404);
    }
    await userRepository.delete(userId);
  },
};

module.exports = userService;
```